System/FilePath/Data/ByteString/Short/Decode.hs

```diff
@@ -0,0 +1,158 @@
+
+import Data.ByteString
+    ( ByteString )
+import Data.ByteString.Short
+    ( ShortByteString )
+import Data.Text
+    ( Text )
+import Data.Text.Encoding.Error
+    ( OnDecodeError, UnicodeException, strictDecode )
+import Data.Text.Internal.Fusion.Types
+    ( Step (..), Stream (..) )
+import Data.Text.Internal.Fusion.Size
+    ( maxSize )
+import Data.Text.Internal.Unsafe.Char
+#if MIN_VERSION_text(2,0,0)
+    ( unsafeChr16, unsafeChr8 )
+#else
+    ( unsafeChr, unsafeChr8 )
+#endif
+import Data.Bits
+    ( unsafeShiftL, unsafeShiftR )
+import Data.Word
+    ( Word16, Word8 )
+
+import Control.Exception
+    ( evaluate, try )
+import qualified Data.ByteString.Short as BS
+    ( index, length )
+import qualified Data.Text.Encoding as E
+import qualified Data.Text.Internal.Encoding.Utf16 as U16
+import qualified Data.Text.Internal.Encoding.Utf8 as U8
+import GHC.IO
+    ( unsafeDupablePerformIO )
+
+#if !MIN_VERSION_text(2,0,0)
+unsafeChr16 :: Word16 -> Char
+unsafeChr16 = unsafeChr
+#endif
+
+-- | /O(n)/ Convert a 'ByteString' into a 'Stream Char', using UTF-8
+-- encoding.
+streamUtf8 :: OnDecodeError -> ShortByteString -> Stream Char
+streamUtf8 onErr bs = Stream next 0 (maxSize l)
+    where
+      l = BS.length bs
+      next i
+          | i >= l = Done
+          | U8.validate1 x1 = Yield (unsafeChr8 x1) (i+1)
+          | i+1 < l && U8.validate2 x1 x2 = Yield (U8.chr2 x1 x2) (i+2)
+          | i+2 < l && U8.validate3 x1 x2 x3 = Yield (U8.chr3 x1 x2 x3) (i+3)
+          | i+3 < l && U8.validate4 x1 x2 x3 x4 = Yield (U8.chr4 x1 x2 x3 x4) (i+4)
+          | otherwise = decodeError "streamUtf8" "UTF-8" onErr (Just x1) (i+1)
+          where
+            x1 = idx i
+            x2 = idx (i + 1)
+            x3 = idx (i + 2)
+            x4 = idx (i + 3)
+            idx = BS.index bs
+{-# INLINE [0] streamUtf8 #-}
+
+-- | /O(n)/ Convert a 'ShortByteString' into a 'Stream Char', using little
+-- endian UTF-16 encoding.
+streamUtf16LE :: OnDecodeError -> ShortByteString -> Stream Char
+streamUtf16LE onErr bs = Stream next 0 (maxSize (l `unsafeShiftR` 1))
+    where
+      l = BS.length bs
+      {-# INLINE next #-}
+      next i
+          | i >= l                         = Done
+          | i+1 < l && U16.validate1 x1    = Yield (unsafeChr16 x1) (i+2)
+          | i+3 < l && U16.validate2 x1 x2 = Yield (U16.chr2 x1 x2) (i+4)
+          | otherwise = decodeError "streamUtf16LE" "UTF-16LE" onErr Nothing (i+1)
+          where
+            x1    = idx i       + (idx (i + 1) `unsafeShiftL` 8)
+            x2    = idx (i + 2) + (idx (i + 3) `unsafeShiftL` 8)
+            idx = fromIntegral . BS.index bs :: Int -> Word16
+{-# INLINE [0] streamUtf16LE #-}
+
+-- | Decode text from little endian UTF-16 encoding.
+decodeUtf16LEWith :: OnDecodeError -> ShortByteString -> String
+decodeUtf16LEWith onErr bs = unstream (streamUtf16LE onErr bs)
```

Comment:
There is probably little point to stream and immediately unstream. You could have produced a `String` directly from `streamUtf8`.